Cargo test ignores edition

## Summary:
Running `cargo test` does use edition 2015 despite edition 2018 being set.
`cargo {check,run}` are fine.

I've created small repo with example, [click me](https://github.com/mati865/cargo-edition-test).

## Details:
#### Using repo linked above.
1. `cargo check` shows no issues:

```
$ cargo check
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Finished dev [unoptimized + debuginfo] target(s) in 2.11s
```

2. `cargo test` builds tests with edition 2015:
```
$ cargo test
   Compiling edition v0.1.0 (file:///tmp/edition)
error[E0432]: unresolved import `crate::edition`
 --> tests/test.rs:2:12
  |
2 | use crate::edition::info;
  |            ^^^^^^^ Maybe a missing `extern crate edition;`?

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: Could not compile `edition`.

To learn more, run the command again with --verbose.
```

3. Adjusting tests to edition 2015 causes whole create to be build with that edition:
```
$ sed 's_//__' -i tests/test.rs
$ cargo test
   Compiling edition v0.1.0 (file:///tmp/edition)
    Finished dev [unoptimized + debuginfo] target(s) in 0.29s
     Running target/debug/deps/edition-0fd02650983f7d39

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/test-69df21cbf8ae95bf

running 1 test
test test_info ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests edition
error[E0432]: unresolved import `log`
 --> /tmp/edition/src/lib.rs:1:5
  |
1 | use log::{info, log};
  |     ^^^ Maybe a missing `extern crate log;`?

error: test failed, to rerun pass '--doc'
```

Sorry, my bad.

Docs are still failing though:
```
   Doc-tests edition
error[E0432]: unresolved import `log`
 --> /tmp/edition/src/lib.rs:1:5
  |
1 | use log::{info, log};
  |     ^^^ Maybe a missing `extern crate log;`?

error: test failed, to rerun pass '--doc'
```
I cannot test it right now but it looks like it should work: https://github.com/rust-lang/cargo/blob/2f7cf4a6db13e7e240b038971fc2d24196608e10/src/cargo/core/compiler/mod.rs#L601-L604

--- a/src/cargo/core/compiler/compilation.rs
+++ b/src/cargo/core/compiler/compilation.rs
@@ -5,7 +5,7 @@ use std::path::PathBuf;
 use semver::Version;
 use lazycell::LazyCell;
 
-use core::{Package, PackageId, Target, TargetKind};
+use core::{Feature, Package, PackageId, Target, TargetKind};
 use util::{self, join_paths, process, CargoResult, Config, ProcessBuilder};
 use super::BuildContext;
 
@@ -93,12 +93,23 @@ impl<'cfg> Compilation<'cfg> {
 
     /// See `process`.
     pub fn rustc_process(&self, pkg: &Package) -> CargoResult<ProcessBuilder> {
-        self.fill_env(self.rustc_process.clone(), pkg, true)
+        let mut p = self.fill_env(self.rustc_process.clone(), pkg, true)?;
+        let manifest = pkg.manifest();
+        if manifest.features().is_enabled(Feature::edition()) {
+            p.arg(format!("--edition={}", manifest.edition()));
+        }
+        Ok(p)
     }
 
     /// See `process`.
     pub fn rustdoc_process(&self, pkg: &Package) -> CargoResult<ProcessBuilder> {
-        self.fill_env(process(&*self.config.rustdoc()?), pkg, false)
+        let mut p = self.fill_env(process(&*self.config.rustdoc()?), pkg, false)?;
+        let manifest = pkg.manifest();
+        if manifest.features().is_enabled(Feature::edition()) {
+            p.arg("-Zunstable-options");
+            p.arg(format!("--edition={}", &manifest.edition()));
+        }
+        Ok(p)
     }
 
     /// See `process`.

--- a/src/cargo/core/compiler/mod.rs
+++ b/src/cargo/core/compiler/mod.rs
@@ -10,7 +10,7 @@ use serde_json;
 
 use core::profiles::{Lto, Profile};
 use core::shell::ColorChoice;
-use core::{Feature, PackageId, Target};
+use core::{PackageId, Target};
 use util::errors::{CargoResult, CargoResultExt, Internal};
 use util::paths;
 use util::{self, machine_message, Freshness, ProcessBuilder};
@@ -596,13 +596,6 @@ fn rustdoc<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoResult
         rustdoc.arg("--cfg").arg(&format!("feature=\"{}\"", feat));
     }
 
-    let manifest = unit.pkg.manifest();
-
-    if manifest.features().is_enabled(Feature::edition()) {
-        rustdoc.arg("-Zunstable-options");
-        rustdoc.arg(format!("--edition={}", &manifest.edition()));
-    }
-
     if let Some(ref args) = bcx.extra_args_for(unit) {
         rustdoc.args(args);
     }
@@ -738,11 +731,6 @@ fn build_base_args<'a, 'cfg>(
         if !cx.used_in_plugin.contains(unit) {
             cmd.arg("-C").arg(format!("panic={}", panic));
         }
-    }
-    let manifest = unit.pkg.manifest();
-
-    if manifest.features().is_enabled(Feature::edition()) {
-        cmd.arg(format!("--edition={}", manifest.edition()));
     }
 
     // Disable LTO for host builds as prefer_dynamic and it are mutually

--- a/tests/testsuite/doc.rs
+++ b/tests/testsuite/doc.rs
@@ -1508,6 +1508,13 @@ fn doc_edition() {
             .with_status(0)
             .with_stderr_contains("[RUNNING] `rustdoc [..]-Zunstable-options --edition=2018[..]"),
     );
+
+    assert_that(
+        p.cargo("test -v").masquerade_as_nightly_cargo(),
+        execs()
+            .with_status(0)
+            .with_stderr_contains("[RUNNING] `rustdoc [..]-Zunstable-options --edition=2018[..]")
+    );
 }
 
 // Tests an issue where depending on different versions of the same crate depending on `cfg`s

--- a/tests/testsuite/package.rs
+++ b/tests/testsuite/package.rs
@@ -1122,14 +1122,10 @@ fn test_edition() {
                 // --edition is still in flux and we're not passing -Zunstable-options
                 // from Cargo so it will probably error. Only partially match the output
                 // until stuff stabilizes
-                .with_stderr_contains(format!("\
-[COMPILING] foo v0.0.1 ({url})
-[RUNNING] `rustc --crate-name foo src[/]lib.rs --crate-type lib \
-        --emit=dep-info,link --edition=2018 -C debuginfo=2 \
-        -C metadata=[..] \
-        --out-dir [..] \
-        -L dependency={dir}[/]target[/]debug[/]deps`
-", dir = p.root().display(), url = p.url())),
+                .with_stderr_contains("\
+[COMPILING] foo v0.0.1 ([..])
+[RUNNING] `rustc [..]--edition=2018 [..]
+"),
     );
 }
 
@@ -1156,14 +1152,10 @@ fn test_edition_missing() {
                 // --edition is still in flux and we're not passing -Zunstable-options
                 // from Cargo so it will probably error. Only partially match the output
                 // until stuff stabilizes
-                .with_stderr_contains(format!("\
-[COMPILING] foo v0.0.1 ({url})
-[RUNNING] `rustc --crate-name foo src[/]lib.rs --crate-type lib \
-        --emit=dep-info,link --edition=2015 -C debuginfo=2 \
-        -C metadata=[..] \
-        --out-dir [..] \
-        -L dependency={dir}[/]target[/]debug[/]deps`
-", dir = p.root().display(), url = p.url())),
+                .with_stderr_contains("\
+[COMPILING] foo v0.0.1 ([..])
+[RUNNING] `rustc [..]--edition=2015 [..]
+"),
     );
 }
 

